### PR TITLE
Create test suite

### PR DESCRIPTION
**Pull request by jecarr**
- Thread (non-webapp) code is moved into its own module
- Added test-suite

### CODE REVIEW

File: tests/threadtests.py [hunk 1]
```diff
@@ -0,0 +1,758 @@
...
+        :param table: The table to check whether data has appeared.
+        :param method_name: The test method calling this method (for logging purposes).
+        :param found_check: The method to determine given a result from the database, if data is found.
+        :param expect_found: If we are checking data is found in the table or not.
+        :param fail_msg: The message to report on test failure.
+        **kwargs should match the kwargs of ThreadDB.get()
+        """
+        # If we don't have a way to do the check, skip this test
+        if not callable(found_check):
+            self.skipTest('%s: Not provided with method to check data is%s in table.' %
```
Comment (KadeMorton) [inline tests/threadtests.py:69]: See if we fail when insufficient args are passed and change if so

File: tests/threadtests.py [hunk 2]
```diff
@@ -0,0 +1,758 @@
...
+        cls.rest_svc_with_limit = RestService(cls.web_svc, cls.reg_svc, cls.data_svc, cls.ml_svc, cls.dao,
+                                              queue_limit=random.randint(1, 20))
+        services.update(rest_svc=cls.rest_svc_with_limit)
+        cls.web_api_with_limit = WebAPI(services=services)
+
+    @classmethod
+    def tearDownClass(cls):
+        """Any tidying-up after all the test methods."""
+        # Delete the database so a new DB file is used in next test-run
+        if os.path.isfile(cls.DB_TEST_FILE):
```
Comment (KadeMorton) [inline tests/threadtests.py:302]: Probably can make this a method outside the classes

File: tests/threadtests.py [hunk 3]
```diff
@@ -0,0 +1,758 @@
...
+        urls_missing = dict(file='title,url\nt1,\nt2,\n')
+        # The test cases paired with expected error messages
+        tests = [(wrong_columns, 'Two columns have not been specified'), (wrong_param, 'Error inserting report(s)'),
+                 (too_many_columns, 'Two columns have not been specified'),
+                 (urls_missing, 'CSV is missing text in at least one row')]
+        for test_data, predicted_msg in tests:
+            # Call the CSV REST endpoint with the malformed data and check the response
+            data = dict(index='insert_csv')
+            data.update(test_data)
+            resp = await self.client.post('/rest', json=data)
```
Comment (KadeMorton) [inline tests/threadtests.py:490]: Can group these lines into one

File: tests/threadtests.py [hunk 4]
```diff
@@ -0,0 +1,758 @@
...
+        resp_attacks = await self.client.post('/rest', json=dict(index='confirmed_attacks', sentence_id=sen_id))
+        resp_context_json = await resp_context.json()
+        resp_attacks_json = await resp_attacks.json()
+        self.assertTrue(resp_context.status < 300, msg='Obtaining sentence data resulted in a non-200 response.')
+        self.assertTrue(resp_attacks.status < 300, msg='Obtaining sentence attack-data resulted in a non-200 response.')
+        self.assertEqual(resp_context_json[0].get('attack_uid'), 'd99999',
+                         msg='Confirmed attack not associated with sentence as expected.')
+        self.assertEqual(resp_attacks_json[0].get('uid'), 'd99999',
+                         msg='Confirmed attack not returned in confirmed attacks for sentence.')
+        self.assertTrue(len(resp_attacks_json) > 0, msg='No confirmed attacks appearing for sentence.')
```
Comment (KadeMorton) [inline tests/threadtests.py:725]: Move this assert above other to prevent IndexError if false

File: tests/threadtests.py [hunk 5]
```diff
@@ -0,0 +1,758 @@
...
+        # Obtain one of the sentence IDs
+        sen_id = sentences[0].get('uid')
+        # Delete the sentence
+        data = dict(index='remove_sentence', sentence_id=sen_id)
+        await self.client.post('/rest', json=data)
+        # Confirm the sentence got deleted
+        new_sentences = await self.db.get('report_sentences', equal=dict(report_uid=report_id),
+                                          order_by_asc=dict(sen_index=1))
+        if len(sentences) - 1 != len(new_sentences) or new_sentences[0].get('uid') == sen_id:
+            self.skipTest('Could not test report rollback as removing a sentence did not work as expected.')
```
Comment (KadeMorton) [inline tests/threadtests.py:746]: Should switch to fail as we are not testing removing a sentence anywhere else

**Review by KadeMorton — CHANGES_REQUESTED** [review]
As discussed, there are improvements which can be made.

File: tests/threadtests.py (hunk 3 above)
Comment (jecarr) [reply tests/threadtests.py:490]: *I incorrectly flagged this - I need the update() call so this will end up being a few lines (as far as I can see)

**Review by KadeMorton — APPROVED** [review]

**Pull request merged**